diff --git a/src/rdb.c b/src/rdb.c
index 470de9806..3ca1f39fd 100644
--- a/src/rdb.c
+++ b/src/rdb.c
@@ -1832,9 +1832,21 @@ static int rdbSaveInternal(int req, const char *filename, rdbSaveInfo *rsi, int
     int saved_errno;
     char *err_op;    /* For a detailed log */
 
-    FILE *fp = fopen(filename,"w");
+    
+    
+    char abs_filename[1024];
+    char *cwdp = getcwd(cwd, 1024);
+    if (cwdp && filename[0] != '/' && filename[0] != '\\' && filename[1] != ':') {
+        snprintf(abs_filename, 1024, "%s\\%s", cwdp, filename);
+    } else {
+        snprintf(abs_filename, 1024, "%s", filename);
+    }
+    FILE *fp = fopen(abs_filename, "w");
+
+
     if (!fp) {
         saved_errno = errno;
+
         char *str_err = strerror(errno);
         char *cwdp = getcwd(cwd,MAXPATHLEN);
         serverLog(LL_WARNING,
diff --git a/src/server.c b/src/server.c
index a284f5784..861a0bde2 100644
--- a/src/server.c
+++ b/src/server.c
@@ -2450,9 +2450,9 @@ void initServerConfig(void) {
 
     /* Double constants initialization */
     R_Zero = 0.0;
-    R_PosInf = 1.0/R_Zero;
-    R_NegInf = -1.0/R_Zero;
-    R_Nan = R_Zero/R_Zero;
+    R_PosInf = INFINITY;
+    R_NegInf = -INFINITY;
+    R_Nan = NAN;
 
     /* Command table -- we initialize it here as it is part of the
      * initial configuration, since command names may be changed via
diff --git a/tests/cluster/run.tcl b/tests/cluster/run.tcl
index 1deb9e223..5f58c8a57 100644
--- a/tests/cluster/run.tcl
+++ b/tests/cluster/run.tcl
@@ -1,3 +1,5 @@
+set override_file [file join [file dirname [info script]] exec_override.tcl]
+if {[file exists $override_file]} { source $override_file } else { source [file join [file dirname [info script]] ../../exec_override.tcl] }
 # Cluster test suite.
 #
 # Copyright (C) 2014-Present, Redis Ltd.
diff --git a/tests/sentinel/run.tcl b/tests/sentinel/run.tcl
index 4f2a65699..4a4413cf9 100644
--- a/tests/sentinel/run.tcl
+++ b/tests/sentinel/run.tcl
@@ -1,3 +1,5 @@
+set override_file [file join [file dirname [info script]] exec_override.tcl]
+if {[file exists $override_file]} { source $override_file } else { source [file join [file dirname [info script]] ../../exec_override.tcl] }
 # Sentinel test suite.
 #
 # Copyright (C) 2014-Present, Redis Ltd.
diff --git a/tests/support/server.tcl b/tests/support/server.tcl
index c4e24aa54..f3ed6a66b 100644
--- a/tests/support/server.tcl
+++ b/tests/support/server.tcl
@@ -16,11 +16,19 @@ set ::global_overrides {}
 set ::tags {}
 set ::valgrind_errors {}
 
+proc cat_file {filename} {
+    if {![file exists $filename]} { return "" }
+    set fd [open $filename r]
+    set data [read $fd]
+    close $fd
+    return $data
+}
+
 proc start_server_error {config_file error} {
     set err {}
     append err "Can't start the Redis server\n"
     append err "CONFIGURATION:\n"
-    append err [exec cat $config_file]
+    append err [cat_file $config_file]
     append err "\nERROR:\n"
     append err [string trim $error]
     send_data_packet $::test_server_fd err $err
@@ -111,8 +119,8 @@ proc kill_server config {
     send_data_packet $::test_server_fd server-killing $pid
     # Node might have been stopped in the test
     # Send SIGCONT before SIGTERM, otherwise shutdown may be slow with ASAN.
-    catch {exec kill -SIGCONT $pid}
-    catch {exec kill $pid}
+    catch {exec taskkill /F /PID $pid}
+    catch {exec taskkill /F /PID $pid}
     if {$::valgrind} {
         set max_wait 120000
     } else {
@@ -123,10 +131,10 @@ proc kill_server config {
 
         if {$wait == $max_wait} {
             puts "Forcing process $pid to crash..."
-            catch {exec kill -SEGV $pid}
+            catch {exec taskkill /F /PID $pid}
         } elseif {$wait >= $max_wait * 2} {
             puts "Forcing process $pid to exit..."
-            catch {exec kill -KILL $pid}
+            catch {exec taskkill /F /PID $pid}
         } elseif {$wait % 1000 == 0} {
             puts "Waiting for process $pid to exit..."
         }
@@ -145,10 +153,13 @@ proc kill_server config {
 }
 
 proc is_alive pid {
-    if {[catch {exec kill -0 $pid} err]} {
+    if {[catch {exec tasklist /FI "PID eq $pid"} err]} {
         return 0
     } else {
-        return 1
+        if {[string match "*$pid*" $err]} {
+            return 1
+        }
+        return 0
     }
 }
 
@@ -422,7 +433,7 @@ proc create_server_config_file {filename config config_lines} {
 }
 
 proc spawn_server {config_file stdout stderr args} {
-    set cmd [list src/redis-server $config_file]
+    if {[info exists ::env(REDIS_BIN_DIR)]} { set cmd [list $::env(REDIS_BIN_DIR)/redis-server.exe $config_file] } else { set cmd [list src/redis-server $config_file] }
     set args {*}$args
     if {[llength $args] > 0} {
         lappend cmd {*}$args
@@ -431,21 +442,21 @@ proc spawn_server {config_file stdout stderr args} {
     if {$::valgrind} {
         set pid [exec valgrind --track-origins=yes --trace-children=yes --suppressions=[pwd]/src/valgrind.sup --show-reachable=no --show-possibly-lost=no --leak-check=full {*}$cmd >> $stdout 2>> $stderr &]
     } elseif ($::stack_logging) {
-        set pid [exec /usr/bin/env MallocStackLogging=1 MallocLogFile=/tmp/malloc_log.txt {*}$cmd >> $stdout 2>> $stderr &]
+        set ::env(MallocStackLogging) 1
+        set ::env(MallocLogFile) "/tmp/malloc_log.txt"
+        set pid [exec {*}$cmd >> $stdout 2>> $stderr &]
     } else {
         # ASAN_OPTIONS environment variable is for address sanitizer. If a test
         # tries to allocate huge memory area and expects allocator to return
         # NULL, address sanitizer throws an error without this setting.
-        set env [list \
-            "ASAN_OPTIONS=allocator_may_return_null=1" \
-            "MSAN_OPTIONS=allocator_may_return_null=1" \
-            "TSAN_OPTIONS=allocator_may_return_null=1,detect_deadlocks=0,suppressions=src/tsan.sup" \
-        ]
-        set pid [exec /usr/bin/env {*}$env {*}$cmd >> $stdout 2>> $stderr &]
+        set ::env(ASAN_OPTIONS) "allocator_may_return_null=1"
+        set ::env(MSAN_OPTIONS) "allocator_may_return_null=1"
+        set ::env(TSAN_OPTIONS) "allocator_may_return_null=1,detect_deadlocks=0,suppressions=src/tsan.sup"
+        set pid [exec {*}$cmd >> $stdout 2>> $stderr &]
     }
 
     if {$::wait_server} {
-        set msg "server started PID: $pid. press any key to continue..."
+        set msg "server started PID: \d+. press any key to continue..."
         puts $msg
         read stdin 1
     }
@@ -461,7 +472,7 @@ proc wait_server_started {config_file stdout pid} {
     set maxiter [expr {120*1000/$checkperiod}] ; # Wait up to 2 minutes.
     set port_busy 0
     while 1 {
-        if {[regexp -- " PID: $pid.*Server initialized" [exec cat $stdout]]} {
+        if {[regexp -- "Server initialized" [cat_file $stdout]]} {
             break
         }
         after $checkperiod
@@ -469,14 +480,14 @@ proc wait_server_started {config_file stdout pid} {
         if {$maxiter == 0} {
             start_server_error $config_file "No PID detected in log $stdout"
             puts "--- LOG CONTENT ---"
-            puts [exec cat $stdout]
+            puts [cat_file $stdout]
             puts "-------------------"
             break
         }
 
         # Check if the port is actually busy and the server failed
         # for this reason.
-        if {[regexp {Failed listening on port} [exec cat $stdout]]} {
+        if {[regexp {Failed listening on port} [cat_file $stdout]]} {
             set port_busy 1
             break
         }
@@ -487,11 +498,11 @@ proc wait_server_started {config_file stdout pid} {
 proc dump_server_log {srv} {
     set pid [dict get $srv "pid"]
     puts "\n===== Start of server log (pid $pid) =====\n"
-    puts [exec cat [dict get $srv "stdout"]]
+    puts [cat_file [dict get $srv "stdout"]]
     puts "===== End of server log (pid $pid) =====\n"
 
     puts "\n===== Start of server stderr log (pid $pid) =====\n"
-    puts [exec cat [dict get $srv "stderr"]]
+    puts [cat_file [dict get $srv "stderr"]]
     puts "===== End of server stderr log (pid $pid) =====\n"
 }
 
@@ -633,7 +644,7 @@ proc start_server {options {code undefined}} {
         return
     }
 
-    set data [split [exec cat "tests/assets/$baseconfig"] "\n"]
+    set data [split [cat_file "tests/assets/$baseconfig"] "\n"]
     set config {}
     if {$::tls} {
         if {$::tls_module} {
@@ -671,8 +682,8 @@ proc start_server {options {code undefined}} {
         dict set config port $port
     }
 
-    set unixsocket [file normalize [format "%s/%s" [dict get $config "dir"] "socket"]]
-    dict set config "unixsocket" $unixsocket
+    # set unixsocket [file normalize [format "%s/%s" [dict get $config "dir"] "socket"]]
+    # dict set config "unixsocket" $unixsocket
 
     # apply overrides from global space and arguments
     foreach {directive arguments} [concat $::global_overrides $overrides] {
@@ -768,7 +779,7 @@ proc start_server {options {code undefined}} {
 
         if {!$serverisup} {
             set err {}
-            append err [exec cat $stdout] "\n" [exec cat $stderr]
+            append err [cat_file $stdout] "\n" [cat_file $stderr]
             start_server_error $config_file $err
             set ::tags [lrange $::tags 0 end-[llength $tags]]
             return
@@ -790,7 +801,7 @@ proc start_server {options {code undefined}} {
     dict set srv "port" $port
     dict set srv "stdout" $stdout
     dict set srv "stderr" $stderr
-    dict set srv "unixsocket" $unixsocket
+    # dict set srv "unixsocket" $unixsocket
     if {$::tls} {
         dict set srv "pport" $pport
     }
@@ -798,7 +809,9 @@ proc start_server {options {code undefined}} {
     # if a block of code is supplied, we wait for the server to become
     # available, create a client object and kill the server afterwards
     if {$code ne "undefined"} {
-        set line [exec head -n1 $stdout]
+        set fp [open $stdout r]
+        set line [gets $fp]
+        close $fp
         if {[string match {*already in use*} $line]} {
             set ::tags [lrange $::tags 0 end-[llength $tags]]
             error_and_quit $config_file $line
@@ -959,3 +972,7 @@ proc restart_server {level wait_ready rotate_logs {reconnect 1} {shutdown sigter
         reconnect $level
     }
 }
+
+
+
+
diff --git a/tests/support/util.tcl b/tests/support/util.tcl
index 6a011380f..73f3d529b 100644
--- a/tests/support/util.tcl
+++ b/tests/support/util.tcl
@@ -189,15 +189,28 @@ proc wait_lazyfree_done r {
 
 # count current log lines in server's stdout
 proc count_log_lines {srv_idx} {
-    set _ [string trim [exec wc -l < [srv $srv_idx stdout]]]
+    set res 0
+    catch {
+        set fd [open [srv $srv_idx stdout] r]
+        while {[gets $fd line] >= 0} {
+            incr res
+        }
+        close $fd
+    }
+    return $res
 }
 
 # returns the number of times a line with that pattern appears in a file
 proc count_message_lines {file pattern} {
     set res 0
-    # exec fails when grep exists with status other than 0 (when the pattern wasn't found)
     catch {
-        set res [string trim [exec grep $pattern $file 2> /dev/null | wc -l]]
+        set fd [open $file r]
+        while {[gets $fd line] >= 0} {
+            if {[regexp $pattern $line]} {
+                incr res
+            }
+        }
+        close $fd
     }
     return $res
 }
diff --git a/tests/test_helper.tcl b/tests/test_helper.tcl
index 2f73df1c4..26574015e 100644
--- a/tests/test_helper.tcl
+++ b/tests/test_helper.tcl
@@ -1,3 +1,5 @@
+set override_file [file join [file dirname [info script]] exec_override.tcl]
+if {[file exists $override_file]} { source $override_file } else { source [file join [file dirname [info script]] .. exec_override.tcl] }
 # Redis test suite.
 #
 # Copyright (C) 2014-Present, Redis Ltd.
@@ -177,7 +179,7 @@ proc reconnect {args} {
     set host [dict get $srv "host"]
     set port [dict get $srv "port"]
     set config [dict get $srv "config"]
-    set client [redis $host $port 0 $::tls]
+    set f [open "C:/Users/samue/repos/redis/reconnect_trace.txt" a]; puts $f "Calling redis $host $port"; close $f; set client [redis $host $port 0 $::tls]; set f [open "C:/Users/samue/repos/redis/reconnect_trace.txt" a]; puts $f "Done calling redis"; close $f
     if {[dict exists $srv "client"]} {
         set old [dict get $srv "client"]
         $old close
@@ -396,10 +398,11 @@ proc read_from_test_client fd {
             gets stdin
         }
     } elseif {$status eq {exception}} {
-        puts "\[[colorstr red $status]\]: $data"
-        kill_clients
-        force_kill_all_servers
-        exit 1
+        set err "\[[colorstr red $status]\]: $data"
+        puts $err
+        lappend ::failed_tests $err
+        set ::active_clients_task($fd) "(EXCEPTION) $data"
+        signal_idle_client $fd
     } elseif {$status eq {testing}} {
         set ::active_clients_task($fd) "(IN PROGRESS) $data"
     } elseif {$status eq {server-spawning}} {
@@ -530,13 +533,23 @@ proc test_client_main server_port {
         set bytes [gets $::test_server_fd]
         set payload [encoding convertfrom utf-8 [read $::test_server_fd $bytes]]
         foreach {cmd data} $payload break
-        if {$cmd eq {run}} {
-            execute_test_file $data
-        } elseif {$cmd eq {run_code}} {
-            foreach {name filename code} $data break
-            execute_test_code $name $filename $code
-        } else {
-            error "Unknown test client command: $cmd"
+        if {[catch {
+            if {$cmd eq {run}} {
+                execute_test_file $data
+            } elseif {$cmd eq {run_code}} {
+                foreach {name filename code} $data break
+                execute_test_code $name $filename $code
+            } else {
+                error "Unknown test client command: $cmd"
+            }
+        } err]} {
+            set estr "Executing test client: $err.\n$::errorInfo"
+            catch {send_data_packet $::test_server_fd exception $estr}
+            if {$cmd eq {run}} {
+                catch {send_data_packet $::test_server_fd done $data}
+            } elseif {$cmd eq {run_code}} {
+                catch {send_data_packet $::test_server_fd done [lindex $data 0]}
+            }
         }
     }
 }


--- a/tests/cluster/tests/includes/init-tests.tcl
+++ b/tests/cluster/tests/includes/init-tests.tcl
@@ -54,6 +54,6 @@ test "Cluster nodes hard reset" {
         R $id config set repl-diskless-load disabled
         R $id config set cluster-announce-hostname ""
         R $id DEBUG DROP-CLUSTER-PACKET-FILTER -1
-        R $id config rewrite
+        catch {R $id config rewrite}
     }
 }
