diff --git a/tests/test_helper.tcl b/tests/test_helper.tcl
index 2f73df1c4..f436182f1 100644
--- a/tests/test_helper.tcl
+++ b/tests/test_helper.tcl
@@ -1,3 +1,5 @@
+set override_file [file join [file dirname [info script]] exec_override.tcl]
+if {[file exists $override_file]} { source $override_file } else { source [file join [file dirname [info script]] .. exec_override.tcl] }
 # Redis test suite.
 #
 # Copyright (C) 2014-Present, Redis Ltd.
@@ -269,8 +271,8 @@ proc run_solo {name code} {
 proc cleanup {} {
     if {!$::quiet} {puts -nonewline "Cleanup: may take some time... "}
     flush stdout
-    catch {exec rm -rf {*}[glob tests/tmp/redis.conf.*]}
-    catch {exec rm -rf {*}[glob tests/tmp/server.*]}
+    catch {exec rm -rf {*}[glob $::tmproot/redis.conf.*]}
+    catch {exec rm -rf {*}[glob $::tmproot/server.*]}
     if {!$::quiet} {puts "OK"}
 }
 
@@ -279,11 +281,22 @@ proc test_server_main {} {
     set tclsh [info nameofexecutable]
     # Open a listening socket, trying different ports in order to find a
     # non busy one.
-    set clientport [find_available_port [expr {$::baseport - 32}] 32]
+    set clientport [expr {$::baseport - 32}]
+    set maxport [expr {$clientport + 1024}]
+    set server_fd ""
+    while {$clientport < $maxport} {
+        if {[catch {set server_fd [socket -server accept_test_clients -myaddr 127.0.0.1 $clientport]} err]} {
+            incr clientport
+        } else {
+            break
+        }
+    }
+    if {$server_fd eq ""} {
+        error "Failed to find an available port for the test server."
+    }
     if {!$::quiet} {
         puts "Starting test server at port $clientport"
     }
-    socket -server accept_test_clients  -myaddr 127.0.0.1 $clientport
 
     # Start the client instances
     set ::clients_pids {}
@@ -396,10 +409,11 @@ proc read_from_test_client fd {
             gets stdin
         }
     } elseif {$status eq {exception}} {
-        puts "\[[colorstr red $status]\]: $data"
-        kill_clients
-        force_kill_all_servers
-        exit 1
+        set err "\[[colorstr red $status]\]: $data"
+        puts $err
+        lappend ::failed_tests $err
+        set ::active_clients_task($fd) "(EXCEPTION) $data"
+        signal_idle_client $fd
     } elseif {$status eq {testing}} {
         set ::active_clients_task($fd) "(IN PROGRESS) $data"
     } elseif {$status eq {server-spawning}} {
@@ -530,13 +544,23 @@ proc test_client_main server_port {
         set bytes [gets $::test_server_fd]
         set payload [encoding convertfrom utf-8 [read $::test_server_fd $bytes]]
         foreach {cmd data} $payload break
-        if {$cmd eq {run}} {
-            execute_test_file $data
-        } elseif {$cmd eq {run_code}} {
-            foreach {name filename code} $data break
-            execute_test_code $name $filename $code
-        } else {
-            error "Unknown test client command: $cmd"
+        if {[catch {
+            if {$cmd eq {run}} {
+                execute_test_file $data
+            } elseif {$cmd eq {run_code}} {
+                foreach {name filename code} $data break
+                execute_test_code $name $filename $code
+            } else {
+                error "Unknown test client command: $cmd"
+            }
+        } err]} {
+            set estr "Executing test client: $err.\n$::errorInfo"
+            catch {send_data_packet $::test_server_fd exception $estr}
+            if {$cmd eq {run}} {
+                catch {send_data_packet $::test_server_fd done $data}
+            } elseif {$cmd eq {run_code}} {
+                catch {send_data_packet $::test_server_fd done [lindex $data 0]}
+            }
         }
     }
 }
